test_plans/barriers_fences.asciidoc

```diff
@@ -0,0 +1,60 @@
+
+== Testing scope
+
+=== Device coverage
+
+All of the tests described below are performed only on the default device that
+is selected on the CTS command line.
+
+== Tests
+
+=== SYCL2020 interface
+
+==== Testing `sycl::atomic_fence` function
+
+The `atomic_fence(memory_order order, memory_scope scope)` function should be checked with the following `scope` values:
+
+* `sycl::memory_scope::sub_group` - to test synchronozation between work-items only in the same work-group
+* `sycl::memory_scope::work_group` - to test synchronozation between work-items in the same work-group and in different work-groups
+* `sycl::memory_scope::device` - to test synchronozation between work-items in the same work-group and in different work-groups
+* `sycl::memory_scope::system` - to test synchronozation between work-items in the same work-group and in different work-groups
+
+Test should check which `memory_scope` values are supported by the device using `info::device::atomic_fence_scope_capabilities`.
+
+The coherence check between work-items with the `atomic_fence` function should be performed  with `sycl::handler::parallel_for`.
+
+The check should be performed using an integer array in following memory types:
+
+* `local memory` and `sycl::local_accessor` - to test synchronozation between work-items in the same work-group
+* `global memory` and `sycl::accessor` - to test synchronozation between work-items in different work-groups
+
+Array should be half the size of the number of tested `work-items`.
+
+A flag variable of type `atomic_ref<bool, memory_order_relaxed, scope> sync_flag` and function `atomic_fence` should be used to cohere `work-items` with each other.
+
+To check all the `work-items` split into two types:
+
+* writing `work-items` - writes the `expected_val` value to `data[i]`
+* reading `work-items` - checks that the value `data[i] == expected_val`
+
+In the `write-item`, `data[i]` should be initialized with `expected_val` value. After initialization, `atomic_fence(memory_order order, memory_scope scope)` should be called with following values of `order` argument:
+
+* `sycl::memory_order::release` - and then the value of `memory_order::acquire` should be used with reading `work-item`
+* `sycl::memory_order::acq_rel` - both the same values in both `work-items`
+* `sycl::memory_order::seq_cst` - both the same values in both `work-items`
+
+Test should check which `memory_order` values are supported by the device using `info::device::atomic_fence_order_capabilities`.
+
+Next, with writing `work-item`, the value of `sync_flag` should be set to `true`.
+
+In the reading `work-item`, the sequence of operations is as follows:
+
+* store `sync_flag` in a local variable until it reaches the value true
+* call `atomic_fence`
+* check that the `data[i]` value is set to `expected_val`
```

Comment:
> However, I think these tests are not valid because they assume a particular forward-progress guarantee among the work-items. Consider that the reading work-item enters the spin loop on sync_flag before the writer. The code above assumes that the writer will eventually run, but I think there is no such guarantee. @Pennycook has been working on the formalization about forward-progress guarantees, so he might have more to comment here.

I agree that this isn't safe.  It might work on a lot of hardware you run these tests on (because any vectorizing implementations that serialize branches would _probably_ choose to execute the first branch before the second one) but strictly speaking the specification does not guarantee that this code will complete.

The only way to guarantee that a SYCL implementation switches between the work-items in a group is to call `group_barrier`.  But we can't use it here, because the fence implied by `group_barrier` would mean we weren't testing `atomic_fence`.

Reply:
@gmlueck Thank you for the clear example to describe the checking algorithm.

I completely agree with you that there is no guarantee that the writing item will run first. But am I correct in guessing that items are executed in parallel execution as described in [4.9.4.2.2. parallel_for invoke](https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_parallel_for_invoke) and thus the writing item will anyway execute the write and atomic_fence with atomic_ref will provide no data raising situation?

Reply:
Thanks for the clarification. I suppose replacing 'while' with 'if' would be correct in this case.

Reply:
Oh, that's a good idea!  Yes, the reader work-item can do something like:

```
bool write_happened = (sync_flag == true);
sycl::atomic_fence(ORDER2, SCOPE);
if (write_happened) {
  if (*data != EXPECTED)
    /* test fails */
}
```

Or even something like:

```
bool write_happened = false;
for (int i = 0;  i < COUNT;  i++) {
  if (sync_flag == true) {
    write_happened = true;
    break;
}
sycl::atomic_fence(ORDER2, SCOPE);
if (write_happened) {
  if (*data != EXPECTED)
    /* test fails */
}
```

Reply:
Thank you for your proposed variants! I preferred the second of the two.